### PR TITLE
Add PandA HDF writer

### PR DESCRIPTION
**Pull request by jsouter**
Adds in HDF writer for PandA, would not merge yet as relies on a few parts of PandABlocks-ioc that are in flux (@evalott100's working on this, namely a FlushNow record, and I think a few of the PVs like FullFileName, FileName etc will change soon).
Largely duplicates code from the AreaDetector HDF writer, happy to try and move things around so that it could be made more generic.
I've written some additional tests against a real detector running off the PandABlocks-ioc, not included here.
Also, are we happy for the PandA implementation to be outside of the epics subdirectory if it's currently using exclusively epics backend signals?

### CODE REVIEW

File: src/ophyd_async/panda/writers/hdf_writer.py [hunk 1]
```diff
@@ -0,0 +1,112 @@
...
+from .panda_hdf import PandaHDF, _HDFDataset, _HDFFile
+
+
+class PandaHDFWriter(DetectorWriter):
+    def __init__(
+        self,
+        hdf: PandaHDF,
+        directory_provider: DirectoryProvider,
+        name_provider: NameProvider,
+        **scalar_datasets_paths: str,
```
Comment (coretl) [inline src/ophyd_async/panda/writers/hdf_writer.py:25]: I think that we should probably check the ":CAPTURE" pv for each dataset, then only emit the `StreamResource` for the dataset if it isn't "No"

Comment (jsouter) [reply src/ophyd_async/panda/writers/hdf_writer.py:25]: Right, that makes sense

**Comment by jsouter** [comment]
Have moved scalar datasets onto the PandaHDF so that the capturing PVs can be connected along with the others. If we don't like the use of setattr/getattr we can also do something like this https://github.com/bluesky/ophyd-async/pull/90/commits/7e38e155a30d37a1862bc76e9a1de559a9c55df7, but maybe overloading children() is ugly too.

File: src/ophyd_async/panda/writers/panda_hdf.py [hunk 2]
```diff
@@ -0,0 +1,73 @@
...
+        )
+        self.flush_now = epics_signal_rw(bool, prefix + ":HDF5:FlushNow")
+        self.scalar_datasets = scalar_datasets
+        for ds_name, ds_path in self.scalar_datasets.items():
+            setattr(
+                self,
+                "capturing_" + ds_name,
+                epics_signal_r(bool, prefix + ":" + ds_path + ":CAPTURE")
+            )
+        super(PandaHDF, self).__init__(name)
```
[the remark this answers is not in the document]
Comment (jsouter) [reply src/ophyd_async/panda/writers/panda_hdf.py]: Ah I see, I've rewritten to take advantage of this (tests are broken for now, will create some dummy PVI objects so can test with a sim and fix this, but I've tested against the real detector and the functionality seems the same.)

[the remark this answers is not in the document]
Comment (jsouter) [reply src/ophyd_async/panda/writers/panda_hdf.py]: May be more Ophyd-like to set the to_capture parameter to be a Dict of connect signals, currently you pass it a nested dict of block and list signal names e.g. {'pcap': ['samples', 'ts_end']} and it finds e.g. ts_end_capture in the samples pvi block from PVI and creates a signal that can be checked during collect_stream_docs

**Comment by mrakitin** [comment]
@jsouter, @coretl, is it OK if we rebase this PR branch on `main`? It looks like it's out-of-date and not mergeable to main as is. @jwlodek and I are planning to use this branch for our NSLS-II work. Thanks!

**Comment by mrakitin** [comment]
Please disregard my request above - we need to address #140 before rebasing...

**Comment by olliesilvester** [comment]
@mrakitin I'm currently working on an updated version of this, see https://github.com/bluesky/ophyd-async/tree/add_hdf_writer. Created a new branch as this PR is from a forked repo

**Comment by olliesilvester** [comment]
Closing in favour of https://github.com/bluesky/ophyd-async/pull/141/files

**Pull request closed without merging**